candle-transformers/src/generation/mod.rs

```diff
@@ -4,32 +4,77 @@ use rand::{distributions::Distribution, SeedableRng};
 pub struct LogitsProcessor {
     rng: rand::rngs::StdRng,
     temperature: Option<f64>,
+    top_p: Option<f64>,
 }
 
 impl LogitsProcessor {
-    pub fn new(seed: u64, temperature: Option<f64>) -> Self {
+    pub fn new(seed: u64, temperature: Option<f64>, top_p: Option<f64>) -> Self {
         Self {
             rng: rand::rngs::StdRng::seed_from_u64(seed),
             temperature,
+            top_p,
         }
     }
 
+    fn sample_argmax(&mut self, logits: Tensor) -> Result<u32> {
+        let logits_v: Vec<f32> = logits.to_vec1()?;
+        let next_token = logits_v
+            .iter()
+            .enumerate()
+            .max_by(|(_, u), (_, v)| u.total_cmp(v))
+            .map(|(i, _)| i as u32)
+            .unwrap();
+        Ok(next_token)
+    }
+
+    fn sample_multi(&mut self, prs: &Vec<f32>) -> Result<u32> {
+        let distr = rand::distributions::WeightedIndex::new(prs).map_err(Error::wrap)?;
+        let next_token = distr.sample(&mut self.rng) as u32;
+        Ok(next_token)
+    }
+
+    fn sample_topp(&mut self, prs: &mut Vec<f32>, top_p: f32) -> Result<u32> {
```

Comment:
What is the intended behavior if top_p is larger than the highest probability in the array? Weight everything equally? Is it what would happen or will we get some error because of an invalid distribution?

Reply:
It will include all tokens necessary for the probability of the selected tokens to be larger than `top_p`.

For example, if `top_p` is 0.9 and probabilities are `[0.8, 0.1, 0.06, 0.04]` the first three will be selected.

As long as `top_p` is higher than 0, which we check before calling this function, at least one of the tokens will be selected, and we won't get an error.

Reply:
Ah I misunderstood what it does, thought it would only select tokens with some minimum probability.